editoast/src/client/roles.rs

```diff
@@ -0,0 +1,136 @@
+use std::collections::HashSet;
+
+use anyhow::{anyhow, bail};
+use clap::{Args, Subcommand};
+use editoast_authz::{
+    authorizer::{StorageDriver, UserInfo},
+    BuiltinRole,
+};
+use editoast_models::DbConnection;
+use itertools::Itertools as _;
+use strum::IntoEnumIterator;
+use tracing::info;
+
+use crate::models::auth::PgAuthDriver;
+
+#[derive(Debug, Subcommand)]
+pub enum RolesCommand {
+    /// Lists the builtin roles supported by editoast
+    ListRoles,
+    /// Lists the roles assigned to a subject
+    List(ListArgs),
+    /// Grants builtin roles to a subject
+    Add(AddArgs),
+    /// Revokes builtin roles from a subject
+    Remove(RemoveArgs),
+}
+
+#[derive(Debug, Args)]
+pub struct ListArgs {
+    /// A subject ID or user identity
+    subject: String,
+}
+
+#[derive(Debug, Args)]
+pub struct AddArgs {
+    /// A subject ID or user identity
+    subject: String,
+    /// A non-empty list of builtin roles
+    roles: Vec<String>,
+}
+
+#[derive(Debug, Args)]
+pub struct RemoveArgs {
+    /// A subject ID or user identity
+    subject: String,
+    /// A non-empty list of builtin roles
+    roles: Vec<String>,
+}
+
+pub fn list_roles() {
+    BuiltinRole::iter().for_each(|role| println!("{role}"));
+}
+
+async fn parse_and_fetch_subject_id(
+    subject: &String,
+    driver: &PgAuthDriver<BuiltinRole>,
+) -> anyhow::Result<i64> {
+    if let Ok(id) = subject.parse::<i64>() {
+        if let Some(UserInfo { identity, name }) = driver.get_user_info(id).await? {
+            info!("{name} ({identity}) found with ID {id}");
+            Ok(id)
+        } else {
+            bail!("No subject found with ID {id}");
+        }
+    } else {
+        let uid = driver.get_user_id(&subject).await?;
+        uid.ok_or_else(|| anyhow!("No subject {subject} found for"))
+    }
+}
+
+pub async fn list_subject_roles(
+    ListArgs { subject }: ListArgs,
+    conn: DbConnection,
+) -> anyhow::Result<()> {
+    let driver = PgAuthDriver::<BuiltinRole>::new(conn);
+    let subject_id = parse_and_fetch_subject_id(&subject, &driver).await?;
+    let roles = driver.fetch_subject_roles(subject_id).await?;
+    if roles.is_empty() {
+        println!("Subject {subject} has no roles assigned");
+        return Ok(());
+    }
+    for role in roles {
+        println!("{role}");
+    }
+    Ok(())
+}
+
+pub async fn add_roles(
+    AddArgs { subject, roles }: AddArgs,
+    conn: DbConnection,
+) -> anyhow::Result<()> {
+    let driver = PgAuthDriver::<BuiltinRole>::new(conn);
+    let subject_id = parse_and_fetch_subject_id(&subject, &driver).await?;
+    let roles = roles
+        .iter()
+        .map(|role| {
+            role.parse::<BuiltinRole>()
```

Comment:
This auto implementation is case sensitive which is not ideal. We could:
- Implement manually `std::str::FromStr` to be case insensitive.
- Add a lowercase version of all variant and lower case the input role.

Reply:
I'm a bit hesitant to add such DoF to the builtin roles serialization at the API level just for this use case.

I've implemented this behavior in the parsing step directly (admittedly it's a O(n^2) operation but it's in the CLI honestly who cares). I'll let you resolve if it's ok.